compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -363,6 +363,24 @@ impl<'tcx> Inliner<'tcx> {
                     return None;
                 }
 
+                // Do not inline blanket impls since they may be specialized by
+                // a `dyn Trait`'s own built-in implementation.
+                if let InstanceDef::Item(impl_item) = callee.def
+                    && let Some(assoc_item) = self.tcx.opt_associated_item(impl_item)
+                {
+                    let self_ty = match assoc_item.container {
+                        ty::AssocItemContainer::ImplContainer => self
+                            .tcx
+                            .type_of(assoc_item.container_id(self.tcx))
+                            .instantiate(self.tcx, callee.args),
+                        ty::AssocItemContainer::TraitContainer => callee.args.type_at(0),
+                    };
+
+                    if matches!(self_ty.kind(), ty::Param(..)) {
```

Comment:
```suggestion
                    if !self_ty.is_rigid() {
```
please not reject the instance of there's a `self_ty: Sized` bound on the impl